### ISSUE TITLE
Back and Forward actions (keyboard)

### ISSUE THREAD

**Issue by khakhalin** [Zettlr/Zettlr#1590]
It seems that this issue was mentioned before, but was closed, still a clear keyboard navigation (that existed in version 1.6.0 and prior) hadn't yet been reimplemented. (Or at least I cannot find it; sorry if I'm just being slow here)

One of the previous threads pertaining to the issue:
https://github.com/Zettlr/Zettlr/issues/240

A brief description of the issue (as observed in version 1.8.4). Say, you are on a note, and a note contains a link `[[note2]]`. You click on this link to navigate to this 2nd note (using Ctrl-Click). You get to that other note. But now there's no way back. Previously there was a sequence that emulated a "back" key (`Ctrl-1`), but now it no longer works. My understanding is that it was repurposed for tabs funcionality, but it would be so nice to be able to work in a Zettlekasten mode without cluttering the screen with multitudes of simultaneiously open tabs! Dive - edit - resurface back.

A full-fledged "Back" shortcut (with a memory of previously visited pages) would have been even better (because it's the most intuitive document-related mindset, people have now), but at least a one-step "back" button seems necessary.

As an inferior alternative, there could be some sort of "advanced click" (like Ctrl-Shift or something) to open a note in a new tab. I would personally dislike that (clutter), but at least it would have some temporary workaround.

(On a personal note, I didn't update from 1.6.0 through the entire 1.7 era, hoping that the funcionality would come back, but so far it didn't. For me means downgading to 1.6.0 once again, as I'm in a middle of a project, and cannot stop using Zettlakasten now; the exit costs are just too high... But it's a crucial piece of functionality! Thank you!! :)

**Comment by nathanlesage** [comment]
Right now, I'm more open-minded to that functionality, but we'd need some free shortcuts. Implementing such a history is super easy in itself. Any ideas for a shortcut? Preferrably something recognnisable for Word users (if Word even has such a thing lol)

**Comment by khakhalin** [comment]
I think the standard in the field (both Chrome and Word follow it) is to use Alt-LeftArrow for "back". In Word it's used for internal within-document links (say, from a table of contents to a title), in Chrome it's obviously used for pages, but the idea is the same.

Is Alt-LeftArrow still vacant in Zettlr? :)

**Comment by nathanlesage** [comment]
> Is Alt-LeftArrow still vacant in Zettlr? :)

Not really sure, does something happen if you press that, while various parts of the app are focused (e.g. the file manager, the editor, the sidebar, etc.) …?

**Comment by cdaven** [comment]
I also miss this feature, and `Alt+Left` seems like a good choice.

In the editor, it seems that `Alt+Left` moves the cursor to the start of the paragraph (not the line).

**Comment by khakhalin** [comment]
I can't test it right now (had to downgrade back to v1.6 to keep working on a project), but how can we proceed from here? Even if there's a shortcut that is seemingly in use (as Christian said), can we agree to supersede it with a more in-demand, community-supported, standard-among-many-other-programs shortcut?

An alternative would be to invent some other weird shortcut, but that's probably less desirable, no?

I also found this funny and very incomplete website that tries to colelct shortcuts among different programs, and "Alt-LeftArrow" seems to be the most popular option in various IDEs as well: https://defkey.com/search?irq=navigate+back

**Comment by derwok** [comment]
Is there any chance you prioritize this old topic?  I get's a whole new need with 2.0's "browser like behavior"

**TL;DR: Without a "browser like back/forward" functionality, the usability of links/cross-references between notes is currently broken, IMHO.**

With Zettlr 2.0 a new "Navigation" behavior was introduced: Whenever I click on a link, the link target's document is not opened in a new Tab, but replacing the document of the current tab. This behaviour is described in the [2.0 release notes](https://github.com/Zettlr/Zettlr/releases/tag/v2.0.0) as _"New Browser-Like Behavior"_, that _"[...] is very closely aligned with how modern browsers handle links."_

While I really like this new behavior (as it avoids too many open tabs), I heavily miss a navigate back and forward functionality - which in turn EVERY browser offers. My notes are heavily linked. And I used links a lot. Now with 2.x - whenever I click a link, there is no way back. I lose the source context. And have to either manually insert back-links, or remember the source document, find it in the side bar, and re-open it in a new tab. Both is cumbersome.

Known workaround: Switch of "Avoid opening new tabs if possible" in the settings.
Which I now "unchecked" with a heavy heart... 💔

**Comment by khakhalin** [comment]
Oh, so this feature was not reincluded in 2.0?? I'm still sitting on 1.6, and had put it in my "todo" for the week to move to 2.0 as I hoped the feature is finally there... I fully agree, without a back-and-forth navigation, using Zettlr in oldschool Zettelkalsten mode is next to impossible!

Can we please, please, have a "back" button (and/or an easy shortcut?) It feels so weird to pay on Patreon and popularize Zettlr on Twitter, for almost 2 years now, and not being able to upgrate to newer versions, just because one of the key (and relatively simple) navigation features was randomly gutted at some point, and never reintroduced...

Please?

**Comment by nathanlesage** [comment]
I agree with both of you, and actually I noticed that it's kind of meh that whenever I create a new file I "lose" the currently open tab. I kind of got used to it by now, but it's certainly improvable, yes!

Right now I'm still lacking time since my first PhD examination will be on Nov 30, but I agree that it would be nice to have that back!

But just to be clear: It was never a feature to be gutted, it was an unintended sideeffect which I got rid of by making the navigation work correctly (as in: as intended).

But I just got an idea of how to implement this technically (please feel free to comment if I overlooked something!):

What I could do is create some kind of "history" of the document tabs, so whenever a file is being closed, replaced, or opened, this could be considered one "state", and whenever one presses some "back" shortcut, we could move around in these states … or do I overlook something?

***

_This is btw one of the reasons why I didn't have the time to re-implement this: A lot of time is actually spent thinking about how to implement this on a technical level. I'm not doing this because I want to lose users, but because I have simply no clue of how to do it. This is basically true for most older issues here on the issue tracker which haven't been closed …_

**Comment by nathanlesage** [comment]
@khakhalin I have now (finally) begun working on a session history for open documents that should then allow you to go back and forth between opened documents. Give me … maybe a few days, it'll then come with 2.1.

**Comment by khakhalin** [comment]
Thank you so much for remembering about it!! I'm sure it will be a beloved feature, at least within the Luhmann-affected crowd :)

**Comment by derwok** [comment]
> I'm still lacking time since my first PhD examination will be on Nov 30

@nathanlesage And - before going back to business - can we applaud to a passed PhD, PLEASE? 🤩

**Comment by khakhalin** [comment]
> @nathanlesage And - before going back to business - can we applaud to a passed PhD, PLEASE? 🤩

Wow, is it Dr. Erz now??? 😱😃🥳 !! ??

**Comment by nathanlesage** [comment]
Ehm … there is no passed PhD!

That will happen in the end of 2024/beginning of 2025. I'm just through the very first milestone. A FIRST examination, but certainly not the last 😉

**Comment by khakhalin** [comment]
OK then; happy dissertation 0.1b release! :) Still a supercool milestone! :)

**Comment by derwok** [comment]
> Ehm … there is no passed PhD!
Ah OK. Then sorry for the misunderstanding. 
So, we'll keep our fingers crossed for 3 more years. No problem!

**Comment by nathanlesage** [comment]
Here we go, finally! Now it's not an "accidental" feature anymore, but a real one. Next step is to implement mouse forward and back buttons, buuuuut that will take time, because there's right now no support by Electron for it.

**Comment by derwok** [comment]
Hey! Thanks for implementing this... 🙏

OTOH: In the changelog I read the shortcuts beeing `Cmd/Ctrl+[` and `Cmd/Ctrl+]` - could we re-consider/discuss this? 

1. On non US keyboards this is probably a shortcut that often does not work. As square brackets need some additional modifier (e.g. German PC: AltGr+ö/ä is square bracket and German Mac: Alt/Option+5/6 is square brackets. in almost all programs from PhotoShop (brush size) to Code Editors (Sublime: Code fold/unfold) these create trouble. I might be wrong, but have you tried on PC / Mac if the shortcut works?

2. In the discussion thread of this issue a different shortcut was suggested... The convention from all browsers I know (Edge, Firefox, Chrome) on whatever Platform I know (Windows, Mac, Linux), the shortcut for navigate back/forward is standardized to `Alt+Left` and `Alt+Right`. Many other programs have adoptet these (e.g. MS Word working on Webcontent).

3. But wait! Its not that easy! Zettlr is NOT just a webbrowser (who serves content in read-only mode), but also an Editor. And so, a problem might be that Mac as a Standard has `Alt/Option+Left/Right` to jump between words in all sorts of Editors (Similar to Ctrl+Left/Right in Win & Linux). This would be a pain to lose! But for Example Jetbrains IDEs - you can "browse" between code files by Cmd/Ctrl clicking on whatever symbols. But similar to Zettlr they need to keep Alt+Left/Right reserved for jump-word-wise for Mac users. They solved this problem by putting Navigation Back / Forward on `Ctrl+Alt/Option+Left`and `Ctrl+Alt/Option+Right`

So again; just my 2 cts.

WOK.

**Comment by khakhalin** [comment]
I haven't had a chance to test the new functionality yet (although I'm super-excited!!!), but for the love of God, please, don't put it on any esoteric Ctrl-Alt-Option-Power-Foot-Sacrifice combo! Alt-keys would be great (Macs default layouts are an abomination anyways, and need to be rewired). Ctrl-[ are ok too (I think most Germans are gradually switching to US layout for serious work, are they not?). Other alternatives I've seen use Ctrl-Backspace for back (i think this was the old web-standard, right?), or Ctrl-1 (old Zettlr combo :)

Doesn't Zettlr have a shortcut editor though? Maybe Ctrl-ö for Germans (or something like that) is also an option?

Either way, super-thanks again for implementing it; I'll be testing (and hopefully, finally, using!) it soon!! :)

**Comment by nathanlesage** [comment]
> Ctrl-Alt-Option-Power-Foot-Sacrifice combo

😂

> Doesn't Zettlr have a shortcut editor though?

It's on the roadmap, but as so many things, it takes time

> but have you tried on PC / Mac if the shortcut works?

Yup, works as expected

> The convention from all browsers I know (Edge, Firefox, Chrome) on whatever Platform I know (Windows, Mac, Linux), the shortcut for navigate back/forward is standardized to Alt+Left and Alt+Right

I actually looked that up, have a look:

![image](https://user-images.githubusercontent.com/17251683/146236944-2de3efca-19f4-45be-bc5d-7edcecb343ab.png)

(Yes, I may currently be writing a piece on log4j)

> But wait! Its not that easy! Zettlr is NOT just a webbrowser (who serves content in read-only mode), but also an Editor. 

Yup, but I gave it a thought and the `Cmd/Ctrl+[` and `Cmd/Ctrl+]` made sense, since it's not yet taken anywhere.

But, I mean, changing the shortcut is pretty easy afterwards. This one "just works" for now, so I guess we can leave it at that and later see if that actually impedes anyone's productivity.

And, if, at some point, we actually get a shortcut editor, this won't be an issue after all!

**Comment by nathanlesage** [comment]
Mouse forward/backward buttons are not supported unfortunately, there is no way for me to listen to those mouse events unfortunately. Should that change, it is my first priority to implement precisely that.

**Comment by Joelius300** [comment]
I really do not mean to offend but I find that hard to believe. Many other electron applications like Discord and VS Code support these buttons and online I was able to find quite a few references about built-in forward and backwards commands. I linked some of them below.

- [Electron api docs for `browser-backward` and `browser-forward` events](https://www.electronjs.org/docs/latest/api/browser-window#event-app-command-windows-linux)
- [Issue in regards to a regression that caused the built-in commands to not work anymore including workaround snippets](https://github.com/electron/electron/issues/17134)
- [VS Code issue about the above regression and how to they approached it](https://github.com/microsoft/vscode/issues/71318)
- [VS Code PR with the fix for the above regression including full support for all OS'](https://github.com/microsoft/vscode/pull/76601)

Again, I do not mean any disrespect and am very grateful for your hard work on the project!

**Comment by nathanlesage** [comment]
@Joelius300 No offense taken, but this information is actually new to me. The last time I looked, it wasn't possible to hook into those events without some serious hacky stuff. Now that it's working (again?) I'm happy to implement.

**Comment by nathanlesage** [comment]
I have now added those events, but they are only supported on Windows/Linux so since I'm currently on my MacBook I was unable to test them. They may or may not work.

**Comment by Joelius300** [comment]
Just tested it on my Manjaro KDE, it works! 🎉 

Thank you so much for the quick response _and_ action!

**Comment by nathanlesage** [comment]
Wonderful! Glad to hear that.
